rust/tools/authorization-logic/raksha_examples/VisionPipelineAuthLogic

```diff
@@ -0,0 +1,153 @@
+
+    // Delegates TPPR the ability to apply tags anywhere
+    "ThirdPartyPrivacyReviewer" canSay hasTag(x, y) :- 
+        isAccessPath(x), isTag(y).
+
+    // Delegate the right to claim that particles are "Local" to TPPR.
+    // (We can think of the principal "LocalDevice" as a group)
+    "ThirdPartyPrivacyReviewer" canSay principalX canActAs "LocalDevice" :-
+        isPrincipal(principalX).
+
+    "LocalDevice" may("do_pure_computation", "raw_video_tag").
+    "LocalDevice" may("egress_to_shopping_service", "product_id_tag").
+
+    // Could also give different permissions to RemoteService for once it
+    // has "product_id_tag".
+
+    // These are rules we should get rid of in the real thing.
+    // The universe relations probably don't need to be decentralized. We could
+    // have rules that populate a "centralized" universe and passes this 
+    // to all the principals. (Or there could be other ways to deal with this).
+    isPrincipal("VideoDataSource").
+    isPrincipal("ImageDetector").
+    isPrincipal("ImageSelector").
+    isPrincipal("ProductIdOutput").
+}
+
+"ThirdPartyPrivacyReviewer" says {
+    "VideoDataSource" canActAs "LocalDevice".
+    "ImageDetector" canActAs "LocalDevice".
+    "ImageSelector" canActAs "LocalDevice".
+    "ProductIdOutput" canActAs "LocalDevice".
+    
+    // TPPR delegates the data transformation to ImageSelector
+    // "ImageSelector" canSay becomes(pathX, "raw_video_tag", "product_id_tag") :-
+    //     isAccessPath(pathX).
+    "ImageSelector" canSay declassifies(pathX, "raw_video_tag") :-
+        hasTag(pathX, "product_id_tag"),   // raw_video_tag is only removed if 
+        objectSelected("EndUser", pathX),  // product_id_tag is added
+        isAccessPath(pathX).
+    "ImageSelector" canSay hasTag(pathX, "product_id_tag") :- 
+        isAccessPath(pathX).
+    "ImageSelector" canSay objectSelected("EndUser", pathX) :-
+        isAccessPath(pathX).
+}
+
+// Maybe we want this part of the policy to sit closer to the particles
+// because sometimes relations will refer to path names ?
+// (We can ostensibly do this as long as we can write
+// `claim <RelationName(...)>` in manifests)
+"VideoDataSource" says "VideoDataSource"
+    will("do_pure_computation", "raw_video_tag").
+"ImageDetector" says "ImageDetecor"
+    will("do_pure_computation", "raw_video_tag").
+"ImageSelector" says {
+    "ImageSelector" will("do_pure_computation", "raw_video_tag").
+    declassifies("is_select_image_id", "raw_video_tag").
+    hasTag("is_select_image_id", "product_id_tag").
+    objectSelected("EndUser", "is_selected_image_id").
+}
+
+// Some queries to test this policy:
+video_is_local = query "EndUser" says   
+    "VideoDataSource" canActAs "LocalDevice"?
+
+local_can_do_computation = query "EndUser" says
+    "LocalDevice" may("do_pure_computation", "raw_video_tag")?
+
+////// These are the queries from the "will" relations that need to be checked
```

Comment:
What is the high-level policy these queries are capturing? Or, is it the case that these are just some test queries?

Reply:
These are not meant to be part of the policy. These are just tests that try to ensure that this part of the code would work if we plug it into the other parts. The `will` statements are meant to be converted into "checks" (not necessarily the arcs keyword) that we can prove the facts that these queries test.

Reply:
The check I'm imagining is the one that the comment towards the bottom tries to describe in natural language. As you pointed out it also fits nicely with how another policy language we have already works, but ties it to principals, delegation, and expands what can be expressed to predicates.